notebooks/agentic_rag_pipeline_with_memory_and_chat.ipynb

```diff
@@ -0,0 +1,530 @@
+{
```

Comment:
<p>I will have a look when possible.</p><p>TBH, I'm not involved with the implementation. I think Julian and Vlad worked on this -&gt; let's get credit to them.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deepset-ai/haystack-cookbook/pull/90/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Reply:
<p>Hi team,</p><p> </p><p>I've just reviewed the notebook, and it seems to be functioning well. However, I have a suggestion to enhance its usability and maintainability. Currently, we're storing chat messages, but as we anticipate handling more complex objects in the future, I recommend implementing a Pydantic-based parser for document content. This approach will streamline object storage in document stores, ensuring type safety and ease of use.</p><p> </p><p>### Proposed Enhancements</p><p> </p><p>1. **Pydantic-Based Parser**:</p><p>&nbsp;&nbsp;- Utilize Pydantic to create a parser for document content. This will allow for type-safe object storage and retrieval.</p><p>&nbsp;&nbsp;- Implementing Pydantic models will make the data validation and serialization processes more robust.</p><p> </p><p>2. **Jinja Templates**:</p><p>&nbsp;&nbsp;- While Jinja templates offer flexibility for specialized results, they are more complex to use.</p><p>&nbsp;&nbsp;- For day-to-day use, we can employ Pydantic for object parsing to simplify operations.</p><p> </p><p>3. **Reader and Writer Components**:</p><p>&nbsp;&nbsp;- Develop additional components: a reader and a writer.</p><p>&nbsp;&nbsp;- These components will use Pydantic serialization to convert documents to objects and vice versa.</p><p>&nbsp;&nbsp;- This approach will maintain type safety and simplify interactions with the document store.</p><p> </p><p>### Benefits</p><p> </p><p>- **Type Safety**: Ensures that the data we store and retrieve adhere to defined schemas, reducing the risk of errors.</p><p>- **Ease of Use**: Simplifies the process of handling documents, making the system more user-friendly.</p><p>- **Future-Proofing**: Prepares the system for managing more complex objects beyond just chat messages.</p><p> </p><p>By implementing these enhancements, we can improve the overall efficiency and reliability of our document handling process. Let me know your thoughts or if there are any other suggestions!</p><p> </p><p>Best regards,</p><p> </p><p>Carlos Fernández Lorán</p>

Reply:
thanks for your feedback @CarlosFerLo. I don't want to add additional complexity to this demo, since the point is to provide a minumum viable / readable example of agentic RAG chat. Definitely for a "production ready" application, adding Pydantic serialization would make sense.